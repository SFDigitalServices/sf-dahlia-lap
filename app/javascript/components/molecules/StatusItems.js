import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import StatusPill from '../atoms/StatusPill'

import StatusItemShape from '../../utils/shapes/StatusItemShape'
import { getSubStatusLabel } from '../../utils/statusUtils'

const StatusDate = ({ timeStamp }) => (
  <div className='status-item-date'>
    {moment.unix(timeStamp).format('MMMM Do YYYY')}
  </div>
)

StatusDate.propTypes = { timeStamp: PropTypes.number }

const StatusItem = ({ statusItem }) => {
  const { status, subStatus, comment, timeStamp } = statusItem

  const subStatusLabel = getSubStatusLabel(status, subStatus)

  return (
    <div className='status-items'>
      <div className='status-item'>
        <div className='status-item-header'>
          <StatusPill status={status} />
          <StatusDate timeStamp={timeStamp} />
        </div>
        {subStatusLabel && (
          <div className='item-substatus-text'>
            {subStatusLabel}
          </div>
        )}
        <div className='status-item-text'>
          {comment}
        </div>
      </div>
    </div>
  )
}

StatusItem.propTypes = { statusItem: PropTypes.shape(StatusItemShape).isRequired }

/**
 * Displays a list of status updates for the Lease-up sidebar.
 *
 * Only displays up to {limit} updates
 *
 * Note that the statusItems prop won't be sorted by this component,
 * the items should be sorted before they're passed in for performance reasons.
 */
const StatusItems = ({ statusItems, limit }) => {
  const limitedItems = statusItems.slice(0, limit)

  return (
    <div className='status-items'>
      {limitedItems.map((item, idx) => (
        <StatusItem
          key={idx}
          statusItem={item}
        />
      ))}
    </div>
  )
}

StatusItems.propTypes = {
  limit: PropTypes.number,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape))
}

StatusItems.defaultProps = {
  limit: Number.MAX_VALUE,
  statusItems: []
}

export default StatusItems
