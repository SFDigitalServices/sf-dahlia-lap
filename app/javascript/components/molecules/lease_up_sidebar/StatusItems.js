import React from 'react'
import PropTypes from 'prop-types'

import StatusItem from './StatusItem'

import StatusItemShape from '../../../utils/shapes/StatusItemShape'

/**
 * Displays a list of status updates for the Lease-up sidebar.
 *
 * Only displays up to {limit} updates
 *
 * Note that the statusItems prop won't be sorted by this component,
 * the items should be sorted before they're passed in for performance reasons.
 */
const StatusItems = ({ statusItems, limit }) => {
  const limitedItems = statusItems.slice(0, limit)

  return (
    <div className='status-items'>
      {limitedItems.map((item, idx) => (
        <StatusItem key={idx} statusItem={item} />
      ))}
    </div>
  )
}

StatusItems.propTypes = {
  limit: PropTypes.number,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape))
}

StatusItems.defaultProps = {
  limit: Number.MAX_VALUE,
  statusItems: []
}

export default StatusItems
