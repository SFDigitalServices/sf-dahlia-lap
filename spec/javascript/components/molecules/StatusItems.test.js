import React from 'react'
import renderer from 'react-test-renderer'
import { mockStatusItem, mockStatusItems } from '../../mocks/statusItemMock'
import StatusItems from 'components/molecules/StatusItems'

const getWrapper = (items, limit = null) => renderer.create(
  <StatusItems
    statusItems={items}
    limit={limit}
  />
)

describe('StatusItems', () => {
  test('should render empty status items correctly', () => {
    const wrapper = getWrapper([])
    expect(wrapper).toMatchSnapshot()
  })

  test('should render a single status item correctly', () => {
    const wrapper = getWrapper([mockStatusItem()])
    expect(wrapper).toMatchSnapshot()
  })

  test('should render multiple status items correctly', () => {
    const wrapper = getWrapper(mockStatusItems())
    expect(wrapper).toMatchSnapshot()
  })

  test('should render multiple status items with limit 0 correctly', () => {
    const wrapper = getWrapper(mockStatusItems(), 0)
    expect(wrapper).toMatchSnapshot()
  })

  test('should render multiple status items with limit 1 correctly', () => {
    const wrapper = getWrapper(mockStatusItems(), 1)
    expect(wrapper).toMatchSnapshot()
  })

  test('should render multiple status items with limit 100 correctly', () => {
    const wrapper = getWrapper(mockStatusItems(), 100)
    expect(wrapper).toMatchSnapshot()
  })
})
