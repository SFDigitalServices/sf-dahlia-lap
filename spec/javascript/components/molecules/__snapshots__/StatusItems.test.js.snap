// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusItems snapshot tests should render a single status item correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Approval letter sent",
        "timeStamp": 1598573280,
      }
    }
  />
</div>
`;

exports[`StatusItems snapshot tests should render empty status items correctly 1`] = `
<div
  className="status-items"
/>
`;

exports[`StatusItems snapshot tests should render multiple status items correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Approval letter sent",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="1"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Unit selected",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="2"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Waiting for subsidy inspection",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="3"
    statusItem={
      Object {
        "comment": null,
        "status": "Waitlisted",
        "subStatus": "Written confirmation sent",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="4"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Lease Signed",
        "subStatus": null,
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="5"
    statusItem={
      Object {
        "comment": null,
        "status": "Lease Signed",
        "subStatus": null,
        "timeStamp": 1598573280,
      }
    }
  />
</div>
`;

exports[`StatusItems snapshot tests should render multiple status items with limit 0 correctly 1`] = `
<div
  className="status-items"
/>
`;

exports[`StatusItems snapshot tests should render multiple status items with limit 1 correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Approval letter sent",
        "timeStamp": 1598573280,
      }
    }
  />
</div>
`;

exports[`StatusItems snapshot tests should render multiple status items with limit 100 correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Approval letter sent",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="1"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Unit selected",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="2"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "subStatus": "Waiting for subsidy inspection",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="3"
    statusItem={
      Object {
        "comment": null,
        "status": "Waitlisted",
        "subStatus": "Written confirmation sent",
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="4"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Lease Signed",
        "subStatus": null,
        "timeStamp": 1598573280,
      }
    }
  />
  <StatusItem
    key="5"
    statusItem={
      Object {
        "comment": null,
        "status": "Lease Signed",
        "subStatus": null,
        "timeStamp": 1598573280,
      }
    }
  />
</div>
`;
